zap/src/test/java/org/zaproxy/zap/extension/pscan/scanner/RegexAutoTagScannerUnitTest.java

```diff
@@ -0,0 +1,118 @@
+import static org.mockito.Mockito.verify;
+
+import net.htmlparser.jericho.Config;
+import net.htmlparser.jericho.Source;
+import org.apache.commons.httpclient.URI;
+import org.apache.commons.httpclient.URIException;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+import org.parosproxy.paros.db.DatabaseException;
+import org.parosproxy.paros.model.HistoryReference;
+import org.parosproxy.paros.network.HttpMalformedHeaderException;
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.zap.ZAP;
+import org.zaproxy.zap.utils.Stats;
+import org.zaproxy.zap.utils.StatsListener;
+
+public class RegexAutoTagScannerUnitTest {
+
+    private static final String BODY =
+            "<html><head>@@head@@</head><body>@@body_one@@ @@body_two@@</body><html>";
+    private static final String MAILTO_PATTERN = "<.*href\\s*=\\s*['\"]?mailto:";
+    private static final String MAILTO_CONFIG = "MailTo";
+    private static final RegexAutoTagScanner MAILTO_TAG_SCAN_RULE =
+            new RegexAutoTagScanner(
+                    MAILTO_CONFIG,
+                    RegexAutoTagScanner.TYPE.TAG,
+                    MAILTO_CONFIG,
+                    null,
+                    null,
+                    null,
+                    MAILTO_PATTERN,
+                    true);
+
+    static {
+        Config.LoggerProvider = ZAP.JERICHO_LOGGER_PROVIDER;
+    }
+
+    private StatsListener listener;
+
+    @BeforeEach
+    public void setUp() {
+        listener = spy(StatsListener.class);
+        Stats.addListener(listener);
+    }
+
+    @AfterEach
+    public void cleanup() {
+        Stats.removeListener(listener);
+        MAILTO_TAG_SCAN_RULE.setEnabled(true);
+    }
+
+    @Test
+    public void shouldNotCountTagWhenBodyDoesNotMatch()
+            throws URIException, HttpMalformedHeaderException, DatabaseException {
+        // Given
+        HttpMessage msg = new HttpMessage(new URI("http://example.com/", true));
+        msg.setHistoryRef(mock(HistoryReference.class));
+        // When
+        MAILTO_TAG_SCAN_RULE.scanHttpResponseReceive(msg, -1, new Source(BODY));
+        // Then
+        Mockito.verifyNoInteractions(listener);
+    }
+
+    @Test
+    public void shouldCountTagWhenBodyHasMatch()
+            throws URIException, HttpMalformedHeaderException, DatabaseException {
+        // Given
+        HttpMessage msg = new HttpMessage(new URI("http://example.com/", true));
+        msg.setResponseBody(BODY.replace("@@body_two@@", "<a href=\"mailto:me@.my.net\">me</a>"));
+        msg.setHistoryRef(mock(HistoryReference.class));
+        // When
+        MAILTO_TAG_SCAN_RULE.scanHttpResponseReceive(msg, -1, new Source(msg.getResponseBody().toString()));
+        // Then
+        verify(listener)
+                .counterInc(
+                        "http://example.com", RegexAutoTagScanner.TAG_STATS_PREFIX + MAILTO_CONFIG);
```

Comment:
This seemed like an easier way to go, since the messages don't actually have a `HistoryReference` from which to get the tags. However, tags are applied and counted from the same method 😀

Reply:
Not always, the counter might be incremented even when not tagged. The test class has the (mocked) `HistoryReference` set previously so it can be verified through that one (or just check the `parent`). These would be another tests anyway.

Reply:
The mocked history reference doesn't get tags. There's no getter for parent and it can't be spy'd because it doesn't have a zero arg construcutor.